src/jit/codegenxarch.cpp

```diff
@@ -7492,22 +7518,27 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
         {
             inst_IV(INS_push, data->gtIntCon.gtIconVal);
         }
+        genStackLevel += argSize;
     }
     else if (data->OperGet() == GT_FIELD_LIST)
     {
         GenTreeFieldList* const fieldList = data->AsFieldList();
         assert(fieldList != nullptr);
 
-        m_pushStkArg      = false;
-        const int argSize = putArgStk->getArgSize();
-        assert((argSize % TARGET_POINTER_SIZE) == 0);
-
-        const bool preAdjustedStack = genAdjustStackForPutArgStk(putArgStk, false);
-
-        // If the stack was not pre-adjusted, set the current field offset to the size of the struct arg (which must be
-        // a multiple of the target pointer size). Otherwise, set the offset to 0.
-        int      currentOffset   = preAdjustedStack ? 0 : argSize;
-        unsigned prevFieldOffset = argSize;
+        // Set m_pushStkArg and pre-adjust the stack if necessary.
+        const bool preAdjustedStack = genAdjustStackForPutArgStk(putArgStk);
+        // For now, we only support the "push" case; we will push a full slot for the first field of each slot
+        // within the struct.
+        assert((putArgStk->isPushKind()) && !preAdjustedStack && m_pushStkArg);
```

Comment:
IMO, GT_PUTARG_STK(GT_FIELD_LIST) is a bit complex to follow and too big to deserve a routine of its own.